### DISCUSSION TITLE
How to use localStorage with SSR?

### DISCUSSION THREAD

**Discussion by onurusluca** [nuxt/nuxt#27793] [Questions] [answered]
I have this small e commerce and I am trying to add a cart functionality that stores cart data to localStorage but as you know one can't simply access localStorage with SSR.

Do you have any suggestions on this?  
What is the best way to do this?

Saving the ids only on cookie and getting them from database is not a solution for me.

Sorry, I don't have a reproduction because I messed up all of my code.
If I find a good solution, I am thinking of opening a PR for the docs and add the needed guide.

Tried:

* [pinia persisted state](https://prazdevs.github.io/pinia-plugin-persistedstate/frameworks/nuxt-3.html)
   * works ok but reactivity is broken? 
   * it uses cookies but cookies has only around 4000bytes storage which is too low for me
   * you can set it to use localStorage but as it says there I can't use it with ssr
* [vueuse useStorage](https://vueuse.org/core/useStorage/)
   * not working? not possible like above? reactivity is broken
* [process.client](https://v2.nuxt.com/docs/internals-glossary/context/)
   * is deprecated in nuxt 3
   * it's [nuxt 3 alternative](https://nuxt.com/docs/api/advanced/import-meta) \`import.meta.client\` is showing if me client or server but how to handle the rest?
* [clientonly](https://nuxt.com/docs/api/components/client-only)
   * tried but i probably did something wrong.
   * is this possible? viable?

used links:

* [https://www.reddit.com/r/Nuxt/comments/1455lit/save\_data\_to\_browser\_localstorage\_from\_an/](https://www.reddit.com/r/Nuxt/comments/1455lit/save_data_to_browser_localstorage_from_an/)
* [https://github.com/nuxt/nuxt/discussions/25500](https://github.com/nuxt/nuxt/discussions/25500)
* [https://stackoverflow.com/questions/77811872/vuejs-nuxt-js-how-to-properly-load-state-from-localstorage-using-uselocalstorag](https://stackoverflow.com/questions/77811872/vuejs-nuxt-js-how-to-properly-load-state-from-localstorage-using-uselocalstorag)
* [https://www.reddit.com/r/Nuxt/comments/14uiy28/hi\_guys\_im\_new\_to\_nuxt\_im\_building\_an\_application/](https://www.reddit.com/r/Nuxt/comments/14uiy28/hi_guys_im_new_to_nuxt_im_building_an_application/)
* [https://lk4real.medium.com/working-with-localstorage-in-nuxt-3-c27618ab89bd](https://lk4real.medium.com/working-with-localstorage-in-nuxt-3-c27618ab89bd)

**Comment by TheAlexLichter** [comment]
If you need access to the localstorage, then your code interacting with it can only be executed on the client (as there is no "browser localstorage" on the server).

You can e.g. wrap your code in `onBeforeMount` or use a "Client-Side" Component (suffix: `.client.vue`)

**Reply by onurusluca** [reply]
Thank you so much for the reply. I am aware of that fact however I am not able to do so.

> wrap your code in onBeforeMount

So in the example of a cart component, I would wrap the localStorage logic in onBeforeMount?

>use a "Client-Side" Component (suffix: .client.vue)

Is there any differences between .client.vue and <ClientOnly> tag?
So I would wrap my cart component in that or do cart.client.vue?
And I can use the localStorage normally?

**Comment by onurusluca** [comment, marked as answer]
Got it resolved! It's best to just store item IDs in the cart and keep the full cart data on the backend. This way, we avoid outdated pricing issues and security risks with users potentially modifying data client-side. Keeping just the cart ID in a cookie simplifies things a lot. Closing this issue now. Thanks!
